### PR TITLE
DLPNO-CCSD(T) Primer Part 1: Reorganize DLPNO Class Heirarchy

### PR DESCRIPTION
**Pull request by andyj10224**
## Description
In preparation of the DLPNO-CCSD(T) code, I have changed the class heirarchy of the DLPNO module. Currently, there is only a `DLPNOMP2` class. However, I have split that class into a `DLPNO -> DLPNOMP2` inheritance heirarchy, to allow for 
`DLPNO -> DLPNOCCSD -> DLPNOCCSD_T` in the future, and potentially `DLPNO -> DLPNOMP2 -> DLPNOMP2_F12`

## Dev notes & details

- [x] Changes class heirarchy of DLPNO module
- [x] Changes the way the DOI grids are handled in DLPNO-MP2, adds separate user options for the DOI grid specifically

## Questions
- [ ] 

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by andyj10224** [comment]
Requesting reviews from @loriab @davpoolechem @jturney

**Review by jturney — APPROVED** [review]

File: psi4/src/psi4/dlpno/dlpno.h [hunk 1]
```diff
@@ -26,21 +26,35 @@
...
+
+#include <map>
+#include <tuple>
+#include <string>
+#include <unordered_map>
 
 namespace psi {
 namespace dlpno {
 
+enum AlgorithmType { MP2, CCSD, CCSD_T };
```
Comment (davpoolechem) [inline psi4/src/psi4/dlpno/dlpno.h:49]: Any reason to use an `enum` here instead of an `enum class`?

File: psi4/src/psi4/dlpno/dlpno.cc [hunk 2]
```diff
@@ -0,0 +1,1243 @@
...
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 2e-2;
+        } else if (options_.get_str("PNO_CONVERGENCE") == "NORMAL") {
+            if (!T_CUT_PNO_changed) T_CUT_PNO_ = 1e-8;
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 1e-2;
+        } else if (options_.get_str("PNO_CONVERGENCE") == "TIGHT") {
+            if (!T_CUT_PNO_changed) T_CUT_PNO_ = 1e-9;
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 5e-3;
+        }
+    } else {
+        throw PSIEXCEPTION("Requested DLPNO algorithm has NOT been implemented yet");
```
Comment (loriab) [inline psi4/src/psi4/dlpno/dlpno.cc:95]: Including the wrapper, this is the third check on the allowed algorithm, plus the enum, plus read_options only lets it be set to MP2. Checks are good, but you might want to reassess the caution/clutter balance once all the code rearrangement settles down.

**Review by loriab — COMMENTED** [review]
Thanks for splitting out a "moving code" PR. After the read_options tweaks, lgtm!

File: psi4/src/psi4/dlpno/dlpno.h (hunk 1 above)
Comment (andyj10224) [reply psi4/src/psi4/dlpno/dlpno.h:49]: Made change to `enum class`

File: psi4/src/read_options.cc [hunk 3]
```diff
@@ -2549,6 +2549,8 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
 
         /*- SUBSECTION Expert Options -*/
 
+        /*- Which DLPNO Algorithm to run (not set by user) !expert*/
```
[the remark this answers is not in the document]
Comment (andyj10224) [reply psi4/src/read_options.cc:2552]: Done

File: psi4/src/read_options.cc [hunk 4]
```diff
@@ -2569,6 +2571,18 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
...
+
+        /*- Number of spherical points in DOI grid !expert -*/
+        options.add_int("DOI_SPHERICAL_POINTS", 50);
+        /*- Number of radial points in DOI grid !expert -*/
+        options.add_int("DOI_RADIAL_POINTS", 25);
+        /*- Screening criteria for basis function values on DOI grids !expert -*/
+        options.add_double("DOI_BASIS_TOLERANCE", 1.0E-10);
+        /*- Pruning scheme for DOI grids !expert -*/
+        options.add_str("DOI_PRUNING_SCHEME", "ROBUST", 
+                        "ROBUST TREUTLER NONE FLAT P_GAUSSIAN D_GAUSSIAN P_SLATER D_SLATER LOG_GAUSSIAN LOG_SLATER NONE");
```
[the remark this answers is not in the document]
Comment (andyj10224) [reply psi4/src/read_options.cc:2585]: Done

File: psi4/src/psi4/dlpno/dlpno.cc (hunk 2 above)
Comment (andyj10224) [reply psi4/src/psi4/dlpno/dlpno.cc:95]: Yes, in this case, this check is redundant

**Comment by andyj10224** [comment]
@loriab @davpoolechem This PR is ready for merge! I am going to rebase it with master and this PR should be merged after all checks pass

**Review by loriab — APPROVED** [review]

**Pull request merged**